VocaDbWeb/Controllers/Api/AlbumApiController.cs

```diff
@@ -249,9 +249,12 @@ public class AlbumApiController : ApiController {
 		[Route("top")]
 		[CacheOutput(ClientTimeSpan = hourInSeconds, ServerTimeSpan = hourInSeconds)]
 		public IEnumerable<AlbumForApiContract> GetTopAlbums(
-			int[] ignoreIds,
+			[FromUri] int[] ignoreIds = null,
 			ContentLanguagePreference languagePreference = ContentLanguagePreference.Default,
-			AlbumOptionalFields fields = AlbumOptionalFields.None) => otherService.GetTopAlbums(languagePreference, fields, ignoreIds ?? new int[0]);
+			AlbumOptionalFields fields = AlbumOptionalFields.None) {
+			ignoreIds ??= Array.Empty<int>();
+			return otherService.GetTopAlbums(languagePreference, fields, Array.Empty<int>()).Where(a => !ignoreIds.Contains(a.Id));

```

Comment:
Since popular albums are cached when the [OtherService.GetTopAlbums](https://github.com/VocaDB/vocadb/blob/846b2b12c07b82cb62054431685fe0bdb9fc6ecd/VocaDbModel/Service/OtherService.cs#L109) method is called, I think this is the right way to filter by IDs. Is it problematic if the cache is invalidated and OtherService.GetTopAlbums is called prior to [OtherService.GetFrontPageContent](https://github.com/VocaDB/vocadb/blob/846b2b12c07b82cb62054431685fe0bdb9fc6ecd/VocaDbModel/Service/OtherService.cs#L450)?

Reply:
Then this list won't include those albums that were filtered out by GetFrontPageContent?

I think the best way would be to split GetTopAlbums method so that it can skip caching. This API method is already cached by the CacheOutput attribute, so additional caching is unnecessary. So, when calling from GetFrontPageContent, it would work as before, but when called from this API method, that "service" level cache is ignored. Another fix would be to include ignoreIds list in the cache key, but that might be problematic if the list is long.

Reply:
> The problem is still that when calling from GetFrontPageContent, it would not filter out any albums.

Why not?

> I think the best way would be to split GetTopAlbums method so that it can skip ObjectCache caching (do not save anything to cache and do not attempt to retrieve from cache). This API method is already cached by the CacheOutput attribute, so additional caching is unnecessary. So, when calling from GetFrontPageContent, it would work as before, but when called from this API method, that "service" level ObjectCache is ignored.

Do you mean we don't need to [filter out by recent albums](https://github.com/VocaDB/vocadb/blob/846b2b12c07b82cb62054431685fe0bdb9fc6ecd/VocaDbModel/Service/OtherService.cs#L449) when calling from the API? I made [another implementation](https://github.com/VocaDB/vocadb/compare/master...ycanardeau:feat/639-fix-album-api-controller-another-impl?expand=1). If we need to filter out those albums, this implementation won't work.

Reply:
> Why not?

I mean, since the list of ignored albums is empty, it gets all albums, caches that and returns that list without filtering, right?

> Do you mean we don't need to [filter out by recent albums](https://github.com/VocaDB/vocadb/blob/846b2b12c07b82cb62054431685fe0bdb9fc6ecd/VocaDbModel/Service/OtherService.cs#L449) when calling from the API? I made [another implementation](https://github.com/VocaDB/vocadb/compare/master...ycanardeau:feat/639-fix-album-api-controller-another-impl?expand=1). If we need to filter out those albums, this implementation won't work.

Of course we should filter out those albums. The link you gave doesn't work for me.

Reply:
> I mean, since the list of ignored albums is empty, it gets all albums, caches that and returns that list without filtering, right?

The list of ignored albums is **not** empty when calling from GetFrontPageContent.

Reply:
Uh, let me try to explain. Maybe I misunderstood.

1. Someone calls this API method.
2. Top albums list gets cached *without* filtering by recent albums.
3. Someone opens the front page, calling GetFrontPageContent.
4. Top albums list is returned from cache, no filtering is done by recent album IDs (which is wrong).

The above process is what is happening currently, to my understanding.

Reply:
Yes, you're right and that's exactly why I asked "Is it problematic if the cache is invalidated and [AlbumApiController.GetTopAlbums](https://github.com/VocaDB/vocadb/blob/846b2b12c07b82cb62054431685fe0bdb9fc6ecd/VocaDbWeb/Controllers/Api/AlbumApiController.cs#L251) is called prior to [OtherService.GetFrontPageContent](https://github.com/VocaDB/vocadb/blob/846b2b12c07b82cb62054431685fe0bdb9fc6ecd/VocaDbModel/Service/OtherService.cs#L450)?". Can we solve this problem by calling GetRecentAlbums from the API method?

Reply:
I suggested fixing that by writing that alternate method which skips the cache. Maybe I should just write the code to explain what I mean. It seems pretty clear to me. But I can't do it now.

Reply:
> Do you mean we don't need to [filter out by recent albums](https://github.com/VocaDB/vocadb/blob/846b2b12c07b82cb62054431685fe0bdb9fc6ecd/VocaDbModel/Service/OtherService.cs#L449) when calling from the API? I made [another implementation](https://github.com/VocaDB/vocadb/compare/master...ycanardeau:feat/639-fix-album-api-controller-another-impl?expand=1). If we need to filter out those albums, this implementation won't work.

I think I wrote the code as you suggested. Why doesn't the link work for you?

Reply:
Ok now I see. Yeah that should work.

>If we need to filter out those albums, this implementation won't work.

Why is that?

Reply:
Ok I think I understand what you mean. So when calling from the API, GetTopAlbums won't automatically filter out recent albums, as it does on the front page. It only filters out those albums in the ignoredIds list. That's how I intended it to work. So, calling from the API controller, it only filters out those albums in the ignoredIds. If you want  to filter out recent albums, you need to add them to the ignoredIds list when calling the API.

Reply:
> So when calling from the API, GetTopAlbums won't automatically filter out recent albums, as it does on the front page. It only filters out those albums in the ignoredIds list.

That's what I mean.

> So, it seems to me that with the "bool cached" flag it works as expected, although I was originally suggesting having two methods and the one with caching can call the "not cached" version, so you wouldn't need to have any of the caching code in that method.

What would you name those two methods?

> I think the best way would be to split GetTopAlbums method so that it can skip ObjectCache caching (do not save anything to cache and do not attempt to retrieve from cache).

We have another problem with this. Since popular albums are [fetched randomly from the database](https://github.com/VocaDB/vocadb/blob/846b2b12c07b82cb62054431685fe0bdb9fc6ecd/VocaDbModel/Service/OtherService.cs#L91), the results will be inconsistent if we skip the ObjectCache caching.

1. Open the front page. Let's say we have `Album 1` and `Album 2`.
1. Call the API method. `Album 3` and `Album 4`.
1. Open the front page. `Album 1` and `Album 2`. (using ObjectCache)
1. Call the API method. `Album 3` and `Album 4`. (using CacheOutput)
1. Wait for one hour.
1. Open the front page. `Album 1` and `Album 2`. (using ObjectCache)
1. Call the API method. `Album 5` and `Album 6`.

The description says `Gets list of top rated albums, same as front page.`.

Reply:
>What would you name those two methods?

I was thinking GetTopAlbumsCached and GetTopAlbums (without cache).

>We have another problem with this. Since popular albums are fetched randomly from the database, the results will be inconsistent if we skip the ObjectCache caching.

I don't see that as problem. What I meant with "same as front page" is that the functionality is the same, but the exact results don't have to be. I would suggest simply changing the description.